### PR TITLE
:feat: add ReferenceArrayField

### PR DESCRIPTION
**Pull request by leesei**
Changes to be committed:
	modified:   docs/Fields.md
	new file:   src/mui/field/ReferenceArrayField.js
	new file:   src/mui/field/ReferenceArrayField.spec.js

### CODE REVIEW

**Comment by leesei** [comment]
fixes #429

File: src/mui/field/ReferenceArrayField.js [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+
+    componentWillReceiveProps(nextProps) {
+        if (this.props.reference !== nextProps.reference ||
+            this.props.record.id !== nextProps.record.id) {
+            this.fetchReferences(nextProps);
+        }
+    }
+
+    fetchReferences({ crudGetOneReference, reference, ids } = this.props) {
+        ids.map(id => crudGetOneReference(reference, id));
```
Comment (fzaninotto) [inline src/mui/field/ReferenceArrayField.js]: Argh! Not good. this will issue n requests no matter what. You should dispatch a single action, probably `crudGetMany`, or create a new one. It's the job of the `restClient` to map that to one or several HTTP requests, according to the abilities of the REST API.

File: src/mui/field/ReferenceArrayField.js [hunk 2]
```diff
@@ -0,0 +1,106 @@
+import React, { Component, PropTypes } from 'react';
+import { connect } from 'react-redux';
+import LinearProgress from 'material-ui/LinearProgress';
+import get from 'lodash.get';
+
+// `crudGetOneReference` uses `crudGetMany` and deboucing
```
Comment (fzaninotto) [inline src/mui/field/ReferenceArrayField.js]: Not good (see below)

File: src/mui/field/ReferenceArrayField.js [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+// its usually more performant on `<List>`
+import { crudGetOneReference as crudGetOneReferenceAction } from '../../actions/dataActions';
+
+/**
+ * A container component that fetches records from another resource specified
+ * by an array of *ids* in current record.
```
Comment (fzaninotto) [inline src/mui/field/ReferenceArrayField.js]: Could you add an example of the expected record structure? Something like:

```
{
  id: 123,
  product_ids: [456, 457, 458],
}
```

File: src/mui/field/ReferenceArrayField.js [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+    resource: PropTypes.string.isRequired,
+    source: PropTypes.string.isRequired,
+};
+
+const mapStateToProps = (state, props) => {
+    const { record, source, reference } = props;
+    const ids = get(record, source);
+    return {
+        ids,
+        data: ids
```
Comment (fzaninotto) [inline src/mui/field/ReferenceArrayField.js]: move that logic to a selector function in the reducer (just like `getIds` for `ReferenceManyField`)

File: docs/Fields.md [hunk 5]
```diff
@@ -377,7 +377,7 @@ Then admin-on-rest renders the `<CommentList>` with a loader for the `<Reference
 
 ## `<ReferenceManyField>`
 
-This component fetches a list of referenced record (using the `GET_MANY_REFERENCE` REST method), and passes the result to an iterator component (like `<SingleFieldList>` or `<Datagrid>`). The iterator component usually has one or more child `<Field>` components.
+This component fetches a list of referenced records by filtering with a field of the current record (using the `GET_MANY_REFERENCE` REST method), and passes the result to an iterator component (like `<SingleFieldList>` or `<Datagrid>`). The iterator component usually has one or more child `<Field>` components.
```
Comment (fzaninotto) [inline docs/Fields.md]: This is not very clear. Could you rephrase, or perhaps illustrate with an example record? Same for `ReferenceArrayField`

**Comment by leesei** [comment]
Updated.
I'm using [prettier/prettier](https://github.com/prettier/prettier) to format my code so there format changes here and there in the modified files..

**Comment by leesei** [comment]
Fixed test case and a bug on rendering before data are not fetched.

File: src/mui/field/ReferenceArrayField.js (hunk 1 above)
Comment (leesei) [inline src/mui/field/ReferenceArrayField.js]: If I use `<ReferenceArrayField>` in a `<List>`, each row will invoke `crudGetMany`.
With `crudGetOneReference`, the action will be debounced and the ids for each rows in the List  will be accumulated in a single `crudGetMany`.
Do we need a new action for the accumulated `crudGetMany`? Or I'm using the wrong action?

Comment (fzaninotto) [inline src/mui/field/ReferenceArrayField.js]: yep, you're right

**Comment by fzaninotto** [comment]
The problem is that I don't use prettier, and your code formatting doesn't fit with the rest of the code (4 spaces for indentation, case in the same line a s switch). Please use the same CS as the rest of the code.

File: src/mui/field/ReferenceArrayField.js (hunk 1 above)
Comment (leesei) [inline src/mui/field/ReferenceArrayField.js]: Right as in we need a new action?

**Comment by fzaninotto** [comment]
Yes

**Comment by leesei** [comment]
Ok, I'll add `crudGetManyReferenceAction` and revert the style changes.

**Comment by leesei** [comment]
Alas, name clash.
`CRUD_GET_ONE_REFERENCE` = many single id debounced to single `GET_MANY`
`CRUD_GET_MANY_REFERENCE` = `GET_MANY_REFERENCE`
`NEW_ACTION` = many id array debounced to single `GET_MANY`
Any suggestion to the name?  

Overloading of "Reference" is causing the issue here.
IMO `CRUD_GET_ONE_REFERENCE` should have been `CRUD_DEBOUNCED_GET_ONE`, `referenceActions.js` should have been `debouncedActions.js`, then the new action could be `CRUD_DEBOUNCED_GET_MANY` in the same file.

**Comment by leesei** [comment]
@fzaninotto ping
Any suggestion to the new action name?
I can use `CRUD_DEBOUNCED_GET_MANY` at the moment and let `CRUD_GET_ONE_REFERENCE` renamed to `CRUD_DEBOUNCED_GET_ONE` in 1.0.

**Comment by wadjeroudi** [comment]
@leesei You should just publish your changes with the name you think is good. (to not block this useful PR). Thx.
I propose CRUD_GET_MANY_BY_ONE_REFERENCE.

**Comment by sherryxiao1988** [comment]
@leesei thanks a lot of writing this PR. I've been waiting for this to get merged. Really need it! Could we get this merged and publish soon? Thanks. @fzaninotto

**Comment by leesei** [comment]
Alas, I may well create a new PR.

**Comment by leesei** [comment]
Closing this in favor of #596.

**Comment by fzaninotto** [comment]
I like the `DEBOUNCE` approach, go for it.
